### DISCUSSION TITLE
Systemvariablen v2

### DISCUSSION THREAD

**Discussion by SukramJ** [SukramJ/aiohomematic#473] [Ideas]
In der CCU gibt es folgende Typen von Systemvariablen:
- Zeichenkette
- Werteliste
- Zahl
- Logikwert
- Alarm

## aktuell:
Die Systemvariablen werden wie folgt dargestellt: 
- `sensor` (Zeichenkette, Werteliste, Zahl)
- `binary_sensor` (Logikwert, Alarm) angelegt.

## geplante Änderungen: 
Die Logik bleibt im wesentlichen für in der CCU als _intern_ markierten Systemvariablen bestehen, wo bei es folgende Ausnahme gibt:
- _Intern_ markierte Wertelisten liefern nicht mehr die **Listenposition** zurück sondern den **Wert**.

Für **nicht** als **intern** gekennzeichnete Systemvariablen gibt es folgende Änderung:
- Wertelisten --> `select` Entität
- Zahl --> `number` Entität
- Logikwert —> `switch` Entität
- Alarm —> `switch` Entität

Durch die Umstellung auf `select`, `number` und `switch` ergeben sich folgende Vorteile:
- Sysvars können direkt ohne zusätzliche Logik in der UI geändert werden. 
- Die allgemeinen Services für `select`, `number` und `switch` können verwendet werden.
- Der Service `homematicip_local.set_variable_value` kann, muss aber nicht mehr verwendet werden.
- Verwendung von Device basierten Automatisierungen.

Für Zeichenketten gibt es keine Änderungen.

Ich bitte um konstruktives Feedback.

**Comment by ThomDietrich** [comment]
Ich finde den Vorschlag gut. Klingt nach einer deutlich besseren Anpassung an HA. 👏

Zur Vollständigkeit deiner Ausführungen: Gehe ich recht in der Annahme dass all diese Entities standardmäßig aktiviert sind?

**Reply by SukramJ** [reply]
Die Entitäten sind nach wie vor standardmäßig deaktiviert.
Vielleicht gibt es hierzu aber noch Feedback.

**Reply by ThomDietrich** [reply]
Würde mich auch interessieren. Ich persönlich befürworte "standardmäßig deaktiviert" 👍

**Comment by SukramJ** [comment]
@Baxxy13 Spricht aus CCU-Sicht etwas gegen die Verwendung des `intern` Kennzeichens  bei Systemvariablen zur Steuerung?

**Reply by Baxxy13** [reply]
Also mit den Bezeichnungen und Optionen innerhalb HA bin ich noch nicht so vertraut, denke aber ich weiß wie es gemeint ist.

Meiner Meinung nach ist der "Internal-Flag" als Indikator aber ungeeignet.
 
Als kleinster gemeinsamer Nenner muss hier CCUx und Derivate betrachtet werden. 
Dort sind "interne SysVars" komplett "unsichtbar", nur in Programmen (WENN / DANN) zeigen die sich und lassen sich auswählen.
Man hat auch keinerlei Optionen (abgesehen von HM-Script, und externen Anwendungen) den Flag zu ändern.
Bei RM ist das kein Problem, hier kann man per Checkbox den Flag setzen sowie zusätzlich die Internen SysVars ein/ausblenden.

Ich hätte aber einen (abgeguckten) Alternativ-Vorschlag der für alle Systeme gleich handhabbar wäre.
Einen speziellen Marker (bspw. -HA-SV-) in der Beschreibung der SysVar. 
Wenn du .DPInfo der SysVar ausliest kommst du da dran, und könntest bei Vorhandensein des Markers entsprechend reagieren.

Der CCU-Jack nutzt das so, da muss MQTT als Marker in der Beschreibung sein damit die SysVar gepublished wird.

Zugegeben nicht sehr elegant, aber recht einfach für alle handhabbar. 

Zu "standardmäßig deaktiviert": Bin ich auch dafür.

**Reply by SukramJ** [reply]
Mir geht es in erster Linie darum das interne Sysvars nicht änderbar (Sensor/Binary_Sensor) in HA angezeigt werden.
Ich denke da in erster Linie an Sysvars die durch Geräte (Wetterstation, Messsteckdosen) erzeugt werden, und die man nicht (einfach) schreiben sollte. Wenn ich das bei mir Richtig gesehen sind die alle als intern geflagt.
Wenn ein Benutzer eine eigene Sysvar anlegt, und kann das flag nicht setzten, weil eine Original-CCU-Derivat verwendet wird, dann ist das vielleicht nicht so schön, aber durchaus akzeptabel.

Zu deinem Alternativvorschlag:
Ich kann dem Vorschlag durchaus was abgewinnen, bzw. finde den mittlerweile besser, aber über den JSON Aufruf `SysVar.getAll `hab ich keinen Zugriff auf die Beschreibung. 
D.h, wenn man das machen will braucht man ein um die Beschreibung erweitertes [SysVar.getAll](https://github.com/eq-3/occu/blob/master/WebUI/www/api/methods/sysvar/getall.tcl) als RegaScript. Das müsstest du aber erstellen @Baxxy13.

@danielperna84 Was denkst du?

**Reply by danielperna84** [reply]
Ich stehe wohl auf dem Schlauch. Welchen Mehrwert erreicht man mit dem genannten Alternativ-Vorschlag? Bzw., welches Problem wird dadurch gelöst? So wie es im Eingangspost beschrieben wurde finde ich es eigentlich gut. Wobei ich der Vollständigkeit halber sagen muss, dass ich nicht mit Variablen arbeite und mir daher nicht der Probleme bewusst bin auf die andere Stoßen.

**Reply by SukramJ** [reply]
Wenn ich das richtig verstanden habe kann das intern Flag kann wohl nur über RaspberryMatic gesetzt werden.

**Reply by Baxxy13** [reply]
Ah, ok. Dann hatte ich das teils falsch interpretiert. 
Wenn ich mich recht erinnere sind die originalen Alarmmeldungen/Servicemeldungen (Zähler) - SysVars intern und alle automatisch angelegten "Zählvariablen" die du erwähnt hast.
Da könnte man den "Internal-Flag" doch als Indikator nehmen. 
Bei diesen SysVars könntest du auch den Namen als Indikator nehmen "svXXXCounter".

Andererseits: 
Wenn ich will (und weiß wie) kann ich die SysVars auch in der WebUI ändern, warum das den HA-Usern vorenthalten.
Wenn "standardmäßig deaktiviert" als default bleibt muss ich mir diese speziellen SysVars ja erstmal aktivieren bevor ich drankomme.

Der Alternativvorschlag bezog sich eher auf die unterschiedliche Behandlung der Werteliste. 
Wenn ich das richtig verstehe, würde...

- eine "interne" Werteliste zum Sensor der den aktuellen Wert ausgibt
- eine "nicht interne" Werteliste zu einer select Entität wo ich die Position (Integer) anwählen kann

Also entweder / oder?

Hmm, das Problem wird sein das Otto-Normal-WebUI-User gar nicht weiß das seine eingegeben Werte tatsächlich nur als "Positionsdaten" abgespeichert und gewählt werden.

**Reply by SukramJ** [reply]
Der Alternativvorschlag wäre auch ein Opt-In, wo man sich bewusst entscheiden muss welche "schreibbaren" Entitäten erstellt werden

**Reply by SukramJ** [reply]
> eine "nicht interne" Werteliste zu einer select Entität wo ich die Position (Integer) anwählen kann

Nein, hier ist der Wert Auswahl. Wenn schon denn schon.

**Reply by SukramJ** [reply]
Ich belass es erstmal beim intern Flag.
Die Umsetzung sollte eigentlich für die meisten Benutzer erst mal passen, ohne das man die flags Ändern muss.

Ich hab mal ein Version (1.8.1.beta1) erstellt, in der man sich die erweiterten Sysvars mal anschauen kann.

**Reply by Baxxy13** [reply]
Sehr gut.

**Reply by Baxxy13** [reply]
> Welchen Mehrwert erreicht man mit dem genannten Alternativ-Vorschlag?

Keinen Mehrwert. Es wäre nur auf allen Systemen "einfacher und gleich" einen "nicht beschreibbar Marker" zu setzen.

> das intern Flag kann wohl nur über RaspberryMatic gesetzt werden

Jein. RaspberryMatic bietet als einziges System die Möglichkeit das per WebUI (Checkbox) sowie Filteroptionen ("systeminterne einblenden") zu handhaben.
Der Rest muss das per Script oder externer Anwendung machen. Hier sind Intern-geflaggte SysVars dann quasi in der ganzen WebUI verschwunden, auch bei Einstellungen > Systemvariablen.

**Reply by SukramJ** [reply]
`Alarmzone 1`, `Anwesenheit`, `HB-RF-ETH Connection` sind bei debmatic nicht als intern geflagt, und werden somit als `switch` erstellt. Ist auf jeden fall bei `HB-RF-ETH Connection` fragwürdig.
Nicht schon, aber ich denke akzeptabel.

**Reply by SukramJ** [reply]
@Baxxy13 hast du ein Script, um das intern-flag einer Sysvar zu ändern?

**Reply by Baxxy13** [reply]
Als Switcher-Script:
```
string sysvar_name = "Heizung_A Modus";
object sysvar = dom.GetObject (ID_SYSTEM_VARIABLES).Get (sysvar_name);
WriteLine("Vorher:\nSysVar: "#sysvar.Name() #" | Intern: "# sysvar.Internal() #" | Sichtbar: "# sysvar.Visible());
if (sysvar.Internal()) { sysvar.Internal(false); }
 else { sysvar.Internal(true); }
WriteLine("Nachher:\nSysVar: "#sysvar.Name() #" | Intern: "# sysvar.Internal() #" | Sichtbar: "# sysvar.Visible());
```
Als Einzeiler (true/false einsetzen):
`dom.GetObject (ID_SYSTEM_VARIABLES).Get ("Heizung_A Modus").Internal(true);`

`Alarmzone 1` und `Anwesenheit` sind Standard-SysVars die jede Zentrale hat, die sind default nie intern.
Zur `HB-RF-ETH Connection` kann ich nichts sagen, habe nur die USB-Platine. 
Die ganzen Watchdog-Alarme von RM werden als Intern erzeugt.

**Reply by SukramJ** [reply]
Danke @Baxxy13 

Nach einer Nacht darüber schlafen und etwas herum probieren mit weitern Systemen musste ich meine Meinung noch mal anpassen:
- da @Baxxy13 's Vorschlag ein Option darstellt, reduziert sich der Breaking Change auf: die Wertelisten-Sysvar gibt jetzt den Listeneintrag statt index zurück.
- weniger Supportaufwand und Verständnisprobleme bei den Usern
- eine für alle Benutzer pflegbare Option, sofern Changelog bzw. Readme gelesen werden.

Ein Sysvar-Script habe ich bereits erstellt.
Als Marker würde ich `hahm` Vorschlagen.

OK?

**Reply by SukramJ** [reply]
Die Variante mit `hahm` als Marker in der Sysvar **Beschreibung** ist in `1.8.2.beta1 `umgesetzt.
